spyder/widgets/recover.py

```diff
@@ -0,0 +1,295 @@
+    """
+    if not data:
+        return _('<i>File name not recorded</i>')
+    res = data['name']
+    try:
+        mtime_as_str = time.strftime('%x %X', time.gmtime(data['mtime']))
+        res += '<br>' + _('Last modified: {}').format(mtime_as_str)
+        res += '<br>' + _('Size: {} bytes').format(data['size'])
+    except KeyError:
+        res += '<br>' + _('<i>File no longer exists</i>')
+    return res
+
+
+def recovery_data_key_function(item):
+    """
+    Function for sorting items in RecoveryDialog.data .
+
+    Sort first by name of the original file, then by name of the autosave file.
+    All items without an original file name will be at the end.
+    """
+    orig_dict, autosave_dict = item
+    if orig_dict:
+        return (0, orig_dict['name'], autosave_dict['name'])
+    else:
+        return (1, 0, autosave_dict['name'])
+
+
+class RecoveryDialog(QDialog):
+    def __init__(self, autosave_dir, autosave_mapping, parent=None):
+        """Constructor."""
+        QDialog.__init__(self, parent)
+        self.layout = QVBoxLayout(self)
+        self.setLayout(self.layout)
+        self.layout.setSpacing(self.layout.spacing() * 3)
+        self.autosave_dir = autosave_dir
+        self.autosave_mapping = autosave_mapping
+        self.files_to_open = []
+        self.gather_data()
+        self.add_label()
+        self.add_grid()
+        self.add_cancel_button()
+        self.setWindowTitle(_('Recover from autosave'))
+
+    def gather_data(self):
+        """
+        Gather data about files which may be recovered.
+
+        The data is stored in self.data as a list of tuples with the data
+        pertaining to the original file and the autosave file. Each element of
+        the tuple is a dict as returned by gather_file_data().
+        """
+        self.data = []
+        try:
+            FileNotFoundError
+        except NameError:  # Python 2
+            FileNotFoundError = OSError
+        # In Python 3, easier to use os.scandir()
+        try:
+            for name in os.listdir(self.autosave_dir):
+                full_name = osp.join(self.autosave_dir, name)
+                if osp.isdir(full_name):
+                    continue
+                for orig, autosave in self.autosave_mapping.items():
+                    if autosave == full_name:
+                        orig_dict = gather_file_data(orig)
+                        break
+                else:
+                    orig_dict = None
+                autosave_dict = gather_file_data(full_name)
+                self.data.append((orig_dict, autosave_dict))
+        except FileNotFoundError:  # autosave dir does not exist
+            pass
+        self.data.sort(key=recovery_data_key_function)
+        self.num_enabled = len(self.data)
+
+    def add_label(self):
+        """Add label with explanation at top of dialog window."""
+        txt = _('Spyder found some autosave files. You can use the autosave '
```

Comment:
Much too wordy and somewhat redundant; also "recover" doesn't actually match the wording of the button ("restore"). I suggest you revise it to simply read something like ``Autosave files found. What would you like to do?\n\nThis will be shown each startup if autosave files are not restored, moved or deleted.`` and then add the following tooltips to each button:

Restore: ``Recover the autosaved file to its original location, replacing the original if it exists.``
Discard: ``Delete the autosaved file.``
Open: ``Display the autosave (and the original, if it exists) in Spyder's Editor, and move or delete it manually.``